design/container-steps.md

```diff
@@ -0,0 +1,55 @@
+# Container steps
+
+Status: proposal
+
+An emerging pattern is running discrete steps of a job in different containers.
+Each one carries along a persistent, read/write volume as its workspace.
+
+An example straight from a customer:
+-	microsoft/dotnet => build C# code
+-	yarn => build and bundle JS
+-	aws/sdk => push static files to S3
+-	gcloud/sdk => download database files from private buckets, update some GCP stuff
+-	docker => build and push image for this app
+-	kubectl => deploy this new image on Kubernetes
+-	our custom container => do checks and send email/Slack
+
+Containers steps can be seen as an alternative to tasks.
+Instead of writing to our custom task.json system and packaging dependencies into a VSIX, you write to Docker's system and package dependencies in an image.
+
+## Considerations
+- We don't (and can't easily) support Alpine Linux containers at the job level.
+There are too many runtime dependencies that it doesn't support.
+But, `docker run`ning into an arbitrary container doesn't require any of that infrastructure.
+So we'd have a great story for using teeny containers and not feel bad about having some extra requirements on full-job containers.
+- Need to support custom volume mappings, environment variables, and custom UIDs.
+
+## Challenges
+- This will be tricky to support on-premises.
+Containers in general rely on access to a Docker registry, and those tend to be cloud-hosted.
+Many Azure DevOps Server instances won't have internet connectivity.
+- When containers write files to volumes, they're written with the UID of the container user.
+This means you either have to be root or running as that UID on the host to access the files.
+This will break things like uploading artifacts unless we run that in a container as root.
+
+## YAML syntax
+
+This is loosely based on `docker` command line syntax.
+
+```yaml
+resources:
+  containers:
+  - image: microsoft/dotnet:latest
+    name: dotnet
+  - image: facebook/yarn:latest
+    name: yarn
+
+steps:
+- run: dotnet
```

Comment:
In the case of a container that is not on a private registry you should be able to
```
- run: microsoft/dotnet:latest
```
we can simply assume if we do not find a container resource with the matching name that this must be an image and we can simply run it.

Reply:
@ericsciple and I talked about that. I wanted to be cautious about overloading the meaning of `run: <string>` with too much magic. If I have a string `- run: python`, that'll pull `python:latest` right up until the day I add a resource called `python`.

We're adding inline resource parsing, though, so we can do:
```yaml
- run:
    image: microsoft/dotnet:latest
```

Let me think about it some more, though - I really like the super abbreviated form.

Reply:
Sure it would change the meeting of -run: python but it would be the result of an explicit action you took.  Given we have no compatibility issues here I think we can push the limits a bit compared to other areas where compatibility with existing features forced a slightly more verbose syntax.

Reply:
Yeah, works for me. I'll update.